polars/src/chunked_array/iterator/mod.rs

```diff
@@ -12,51 +12,113 @@ use std::slice::Iter;
 impl<'a, T> ExactSizeIterator for NumIterManyChunk<'a, T> where T: PolarsNumericType {}
-impl<'a> ExactSizeIterator for Utf8IterSingleChunkReturnUnwrapped<'a> {}
-impl<'a> ExactSizeIterator for Utf8IterManyChunkReturnUnwrapped<'a> {}
-impl<'a> ExactSizeIterator for Utf8IterSingleChunkReturnOption<'a> {}
-impl<'a> ExactSizeIterator for Utf8IterSingleChunkNullCheckReturnOption<'a> {}
-impl<'a> ExactSizeIterator for Utf8IterManyChunkReturnOption<'a> {}
-impl<'a> ExactSizeIterator for Utf8IterManyChunkNullCheckReturnOption<'a> {}
+impl<'a, T> PolarsIterator for NumIterSingleChunk<'a, T> where T: PolarsNumericType {}
+impl<'a, T> PolarsIterator for NumIterSingleChunkNullCheck<'a, T> where T: PolarsNumericType {}
+impl<'a, T> PolarsIterator for NumIterManyChunkNullCheck<'a, T> where T: PolarsNumericType {}
+impl<'a, T> PolarsIterator for NumIterManyChunk<'a, T> where T: PolarsNumericType {}
+
+/// A `PolarsIterator` is an iterator over a `ChunkedArray` which contains polars types. A `PolarsIterator`
+/// must implement `ExactSizeIterator` and `DoubleEndedIterator`.
+pub trait PolarsIterator: ExactSizeIterator + DoubleEndedIterator {}
 
 /// Trait for ChunkedArrays that don't have null values.
+/// The result is the most efficient implementation `Iterator`, according to the number of chunks.
 pub trait IntoNoNullIterator {
     type Item;
     type IntoIter: Iterator<Item = Self::Item>;
 
     fn into_no_null_iter(self) -> Self::IntoIter;
 }
 
-impl<'a, T> IntoNoNullIterator for &'a ChunkedArray<T>
+/// Wrapper strunct to convert an iterator of type `T` into one of type `Option<T>`.  It is useful to make the
+/// `IntoIterator` trait, in which every iterator shall return an `Option<T>`.
+struct SomeIterator<I>(I)
+where
+    I: Iterator;
+
+impl<I> Iterator for SomeIterator<I>
+where
+    I: Iterator,
+{
+    type Item = Option<I::Item>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.0.next().map(Some)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.0.size_hint()
+    }
+}
+
+impl<I> DoubleEndedIterator for SomeIterator<I>
+where
+    I: DoubleEndedIterator,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.0.next_back().map(Some)
+    }
+}
+
+impl<I> ExactSizeIterator for SomeIterator<I> where I: ExactSizeIterator {}
+impl<I> PolarsIterator for SomeIterator<I> where I: PolarsIterator {}
+
+/// Iterator for chunked arrays with just one chunk.
+/// The chunk cannot have null values so it does NOT perform null checks.
+///
+/// The return type is `PolarsNumeriType::Native`.
+pub struct NumIterSingleChunk<'a, T>
+where
+    T: PolarsNumericType,
+{
+    iter: Copied<Iter<'a, T::Native>>,
+}
+
+impl<'a, T> NumIterSingleChunk<'a, T>
+where
+    T: PolarsNumericType,
+{
+    fn new(ca: &'a ChunkedArray<T>) -> Self {
+        let chunk = ca.downcast_chunks()[0];
+        let slice = chunk.value_slice(0, chunk.len());
```

Comment:
After changing the code, the tests started failing because it was getting random values from the iterator, probably because it was getting values out of his range, so it was getting uninitialized memory. Then, I took a look at arrow code to check how the function `value_slice` was written. The code I saw is the following:
```rust
    /// Returns a raw pointer to the values of this array.
    pub fn raw_values(&self) -> *const T::Native {
        unsafe { self.raw_values.get().add(self.data.offset()) }
    }

    /// Returns a slice for the given offset and length
    ///
    /// Note this doesn't do any bound checking, for performance reason.
    pub fn value_slice(&self, offset: usize, len: usize) -> &[T::Native] {
        let raw =
            unsafe { std::slice::from_raw_parts(self.raw_values().add(offset), len) };
        &raw[..]
    }
```
If we observe the function `raw_values`, which is the one used in `value_slice`, it already applies the offset `self.data.offset()`. Then, I can conclude that the function `value_slice` does not need the offset but the index, as it already applies the offset when retrieving the data.

So I will leave this part as it was, as I think it is ok. And I will fix the typos in the comments.

Reply:
You are right. Good catch!